sdk/core/System.ClientModel/src/Pipeline/PipelineTransport.cs

```diff
@@ -1,97 +1,146 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.ClientModel.Internal;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace System.ClientModel.Primitives;
 
 public abstract class PipelineTransport : PipelinePolicy
 {
+    #region CreateMessage
+
     /// <summary>
     /// TBD: needed for inheritdoc.
     /// </summary>
-    /// <param name="message"></param>
-    public void Process(PipelineMessage message)
+    public PipelineMessage CreateMessage()
     {
-        ProcessCore(message);
+        PipelineMessage message = CreateMessageCore();
 
-        if (message.Response is null)
+        if (message.Request is null)
         {
-            throw new InvalidOperationException("Response was not set by transport.");
+            throw new InvalidOperationException("Request was not set on message.");
         }
 
-        message.Response.SetIsError(ClassifyResponse(message));
+        if (message.Response is not null)
+        {
+            throw new InvalidOperationException("Response should not be set before transport is invoked.");
+        }
+
+        return message;
     }
 
+    protected abstract PipelineMessage CreateMessageCore();
+
+    #endregion
+
+    #region Process message
+
+    /// <summary>
+    /// TBD: needed for inheritdoc.
+    /// </summary>
+    /// <param name="message"></param>
+    public void Process(PipelineMessage message)
+        => ProcessSyncOrAsync(message, async: false).EnsureCompleted();
+
     /// <summary>
     /// TBD: needed for inheritdoc.
     /// </summary>
     /// <param name="message"></param>
     public async ValueTask ProcessAsync(PipelineMessage message)
+        => await ProcessSyncOrAsync(message, async: true).ConfigureAwait(false);
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, bool async)
     {
-        await ProcessCoreAsync(message).ConfigureAwait(false);
+        Debug.Assert(message.NetworkTimeout is not null, "NetworkTimeout is not set on PipelineMessage.");
+
+        // Implement network timeout behavior around call to ProcessCore.
+        TimeSpan networkTimeout = (TimeSpan)message.NetworkTimeout!;
+        CancellationToken userToken = message.CancellationToken;
```

Comment:
nit: Is `messageToken` more accurate, given that this could either be the default or a user provided value?
```suggestion
        CancellationToken messageToken = message.CancellationToken;
```

Reply:
Yeah, that's fair.  I was thinking of it as what either the client-author or end-user might have provided (as I was working to understand the logic we have), but you are right it could also be a default.